File: backend_samples/library/dotnet/AzureAIVisionFaceDeviceAttestation/src/Android/PlayIntegrityApi.cs
using System.Diagnostics;
using System.Net.Http;
using Google;
using Google.Apis.Auth.OAuth2;
using Google.Apis.PlayIntegrity.v1;
using Google.Apis.PlayIntegrity.v1.Data;
using Google.Apis.Services;
using Azure.AI.Vision.Face.DeviceAttestation.Configuration;
using Azure.AI.Vision.Face.DeviceAttestation.Logging;
using Azure.AI.Vision.Face.DeviceAttestation.Models;

namespace Azure.AI.Vision.Face.DeviceAttestation.Android;

/// <summary>
/// Outcome of a Play Integrity decode attempt. <see cref="Tolerable"/> is true only
/// for the Google-unavailable (network / HTTP 5xx) or quota-exceeded (HTTP 429)
/// conditions the <c>AllowAndroidAttestationWhenGoogleUnavailable</c> policy may
/// accept; every other failure fails closed.
/// </summary>
internal sealed class IntegrityVerdictResult
{
    public bool Ok { get; private init; }
    public PlayIntegrityVerdict? Verdict { get; private init; }
    public bool Tolerable { get; private init; }
    public string Reason { get; private init; } = "";

    public static IntegrityVerdictResult Success(PlayIntegrityVerdict verdict) => new() { Ok = true, Verdict = verdict, Reason = "ok" };
    public static IntegrityVerdictResult FailTolerable(string reason) => new() { Tolerable = true, Reason = reason };
    public static IntegrityVerdictResult FailHard(string reason) => new() { Reason = reason };
}

/// <summary>
/// The network/credentials boundary for Google's Play Integrity API. Reads the
/// service-account JSON + package name from config and decodes a client-supplied
/// integrity token into an <see cref="IntegrityVerdictResult"/>.
/// </summary>
internal static class PlayIntegrityApi
{
    private const string Scope = "https://www.googleapis.com/auth/playintegrity";

    public static async Task<IntegrityVerdictResult> DecryptAndVerifyIntegrityVerdictAsync(
        AttestationConfig config,
        IAttestationLogger logger,
        string integrityToken)
    {
        var serviceAccountJson = config.GoogleServiceAccountJson;
        if (string.IsNullOrEmpty(serviceAccountJson))
        {
            logger.TrackEvent("AndroidAuth.PlayIntegrity.ConfigMissing", new Dictionary<string, object?> { ["missing"] = "GOOGLE_SERVICE_ACCOUNT_JSON" });
            return IntegrityVerdictResult.FailHard("service_account_not_configured");
        }
        var packageName = config.AndroidPackageName;
        if (string.IsNullOrEmpty(packageName))
        {
            logger.TrackEvent("AndroidAuth.PlayIntegrity.ConfigMissing", new Dictionary<string, object?> { ["missing"] = "ANDROID_PACKAGE_NAME" });
            return IntegrityVerdictResult.FailHard("package_name_not_configured");
        }

        GoogleCredential credential;
        try
        {
            credential = CredentialFactory
                .FromJson<ServiceAccountCredential>(serviceAccountJson)
                .ToGoogleCredential()
                .CreateScoped(Scope);
        }
        catch (Exception e)
        {
            logger.TrackException(e, new Dictionary<string, object?> { ["source"] = "decryptAndVerifyIntegrityVerdict.parseCreds" });
            return IntegrityVerdictResult.FailHard("service_account_parse_error");
        }

        var sw = Stopwatch.StartNew();
        try
        {
            using var service = new PlayIntegrityService(new BaseClientService.Initializer
            {
                HttpClientInitializer = credential,
                ApplicationName = "AzureAIVisionFaceDeviceAttestation",
            });
            var request = service.V1.DecodeIntegrityToken(new DecodeIntegrityTokenRequest { IntegrityToken = integrityToken }, packageName);
            var response = await request.ExecuteAsync();
            sw.Stop();

            if (response?.TokenPayloadExternal is null)
            {
                logger.TrackDependency(Dependency(packageName, sw.Elapsed, success: false, "empty-response"));
                return IntegrityVerdictResult.FailHard("empty_response");
            }

            var verdict = Map(response.TokenPayloadExternal);
            logger.TrackDependency(Dependency(packageName, sw.Elapsed, success: true, "200"));
            return IntegrityVerdictResult.Success(verdict);
        }
        catch (Exception e)
        {
            sw.Stop();
            logger.TrackDependency(Dependency(packageName, sw.Elapsed, success: false, "exception"));
            logger.TrackException(e, new Dictionary<string, object?> { ["source"] = "decryptAndVerifyIntegrityVerdict", ["packageName"] = packageName });
            var kind = ClassifyUnavailable(e);
            return kind is not null ? IntegrityVerdictResult.FailTolerable(kind) : IntegrityVerdictResult.FailHard("api_error");
        }
    }

    /// <summary>
    /// Classify a decode error as a tolerable Google-unavailability condition
    /// ("server_unavailable" for a network error / HTTP 5xx, "quota_exceeded" for
    /// HTTP 429) or null when it must fail closed.
    /// </summary>
    private static string? ClassifyUnavailable(Exception e)
    {
        if (e is GoogleApiException gae)
        {
            int status = (int)gae.HttpStatusCode;
            if (status == 429) return "quota_exceeded";
            if (status >= 500 && status <= 599) return "server_unavailable";
            return null;
        }
        if (e is HttpRequestException) return "server_unavailable";
        if (e is TaskCanceledException || e is OperationCanceledException || e is TimeoutException) return "server_unavailable";
        return null;
    }

    private static DependencyTelemetry Dependency(string packageName, TimeSpan duration, bool success, string resultCode) => new()
    {
        Name = "PlayIntegrity.decodeIntegrityToken",
        Target = "playintegrity.googleapis.com",
        Data = $"packageName={packageName}",
        Duration = duration,
        Success = success,
        ResultCode = resultCode,
        Properties = new Dictionary<string, object?> { ["packageName"] = packageName },
    };

    private static PlayIntegrityVerdict Map(TokenPayloadExternal t) => new()
    {
        RequestDetails = t.RequestDetails is null ? null : new PlayIntegrityRequestDetails
        {
            RequestPackageName = t.RequestDetails.RequestPackageName,
            TimestampMillis = t.RequestDetails.TimestampMillis?.ToString(),
            RequestHash = t.RequestDetails.RequestHash,
        },
        AccountDetails = t.AccountDetails is null ? null : new PlayIntegrityAccountDetails
        {
            AppLicensingVerdict = t.AccountDetails.AppLicensingVerdict,
        },
        AppIntegrity = t.AppIntegrity is null ? null : new PlayIntegrityAppIntegrity
        {
            AppRecognitionVerdict = t.AppIntegrity.AppRecognitionVerdict,
            PackageName = t.AppIntegrity.PackageName,
            CertificateSha256Digest = t.AppIntegrity.CertificateSha256Digest?.ToList(),
            VersionCode = t.AppIntegrity.VersionCode?.ToString(),
        },
        DeviceIntegrity = t.DeviceIntegrity is null ? null : new PlayIntegrityDeviceIntegrity
        {
            DeviceRecognitionVerdict = t.DeviceIntegrity.DeviceRecognitionVerdict?.ToList(),
            RecentDeviceActivity = t.DeviceIntegrity.RecentDeviceActivity is null ? null : new PlayIntegrityRecentDeviceActivity
            {
                DeviceActivityLevel = t.DeviceIntegrity.RecentDeviceActivity.DeviceActivityLevel,
            },
        },
        EnvironmentDetails = t.EnvironmentDetails is null ? null : new PlayIntegrityEnvironmentDetails
        {
            PlayProtectVerdict = t.EnvironmentDetails.PlayProtectVerdict,
            AppAccessRiskVerdict = t.EnvironmentDetails.AppAccessRiskVerdict is null ? null : new PlayIntegrityAppAccessRiskVerdict
            {
                AppsDetected = t.EnvironmentDetails.AppAccessRiskVerdict.AppsDetected?.ToList(),
            },
        },
    };
}

File: backend_samples/library/javascript/AzureAIVisionFaceDeviceAttestation/src/android/play_integrity_api.ts
/**
 * Call Google's Play Integrity API to decode a client-supplied integrity
 * token into a verdict. Reads service-account credentials from
 * `GOOGLE_SERVICE_ACCOUNT_JSON` and the bundle id from `ANDROID_PACKAGE_NAME`.
 *
 * The semantic checks on the returned verdict (requestHash binding, timestamp
 * freshness, app + device integrity) live in `integrity_checks.ts`; this file
 * is just the network/credentials boundary.
 */
import { auth, playintegrity } from '@googleapis/playintegrity';
import {
  trackEvent,
  trackException,
  trackDependency,
} from '../logging';
import { getAttestationConfig } from '../config';
import type { PlayIntegrityVerdict } from './types';

/**
 * Outcome of a Play Integrity decode attempt.
 *
 * `ok: false` with `tolerable: true` is reserved for the two conditions the
 * `allowAndroidAttestationWhenGoogleUnavailable` policy may accept: the Play
 * Integrity API was unreachable (network error / HTTP 5xx) or the decode quota
 * was exhausted (HTTP 429). Every other failure — missing/invalid credentials,
 * wrong package, an empty or rejected token — is `tolerable: false` and must
 * always fail closed.
 */
export type IntegrityVerdictResult =
  | { ok: true; verdict: PlayIntegrityVerdict }
  | { ok: false; tolerable: boolean; reason: string };

/**
 * Classify a decode error as a tolerable Google-unavailability condition
 * (`server_unavailable` for a network error / HTTP 5xx, `quota_exceeded` for
 * HTTP 429) or `null` when it is a hard failure that must fail closed.
 */
function classifyUnavailable(
  error: unknown,
): 'server_unavailable' | 'quota_exceeded' | null {
  if (!error || typeof error !== 'object') return null;
  const e = error as {
    code?: string | number;
    status?: number;
    response?: { status?: number };
    message?: string;
  };
  const httpStatus =
    typeof e.response?.status === 'number'
      ? e.response.status
      : typeof e.status === 'number'
        ? e.status
        : typeof e.code === 'number'
          ? e.code
          : undefined;

  // Quota / rate limit exceeded → HTTP 429 (RESOURCE_EXHAUSTED).
  if (httpStatus === 429) return 'quota_exceeded';

  // Google server-side error → HTTP 5xx.
  if (httpStatus !== undefined && httpStatus >= 500 && httpStatus <= 599) {
    return 'server_unavailable';
  }

  // No HTTP response → network / connectivity failure reaching Google.
  if (httpStatus === undefined) {
    const code = typeof e.code === 'string' ? e.code : '';
    const NETWORK_ERROR_CODES = [
      'ECONNREFUSED', 'ECONNRESET', 'ECONNABORTED', 'ETIMEDOUT',
      'ENOTFOUND', 'EAI_AGAIN', 'EPIPE', 'ENETUNREACH', 'EHOSTUNREACH',
    ];
    if (NETWORK_ERROR_CODES.includes(code)) return 'server_unavailable';
    const msg = (e.message || '').toLowerCase();
    if (msg.includes('timeout') || msg.includes('network') || msg.includes('socket hang up')) {
      return 'server_unavailable';
    }
  }

  return null;
}

/**
 * Decrypt and verify a Google Play Integrity token into a verdict.
 *
 * Returns a discriminated {@link IntegrityVerdictResult}: a verdict on success,
 * or `{ ok: false, tolerable }` on failure, where `tolerable` marks the
 * Google-unavailable / quota-exceeded conditions the caller may accept when
 * `allowAndroidAttestationWhenGoogleUnavailable` is set.
 *
 * @param integrityToken - The integrity token from the client
 */
export async function decryptAndVerifyIntegrityVerdict(
  integrityToken: string,
): Promise<IntegrityVerdictResult> {
  // Get service account credentials from environment variable
  const serviceAccountJson = getAttestationConfig().googleServiceAccountJson;
  if (!serviceAccountJson) {
    trackEvent('AndroidAuth.PlayIntegrity.ConfigMissing', {
      missing: 'GOOGLE_SERVICE_ACCOUNT_JSON',
    });
    return { ok: false, tolerable: false, reason: 'service_account_not_configured' };
  }

  // Parse service account credentials
  let credentials;
  try {
    credentials = JSON.parse(serviceAccountJson);
  } catch (error) {
    trackException(error, { source: 'decryptAndVerifyIntegrityVerdict.parseCreds' });
    return { ok: false, tolerable: false, reason: 'service_account_parse_error' };
  }

  // Use the Android package name for the liveness app
  const packageName = getAttestationConfig().androidPackageName;
  if (!packageName) {
    trackEvent('AndroidAuth.PlayIntegrity.ConfigMissing', {
      missing: 'ANDROID_PACKAGE_NAME',
    });
    return { ok: false, tolerable: false, reason: 'package_name_not_configured' };
  }

  const depStart = Date.now();
  try {
    // Create auth client with service account
    const authClient = new auth.GoogleAuth({
      credentials,
      scopes: ['https://www.googleapis.com/auth/playintegrity'],
    });

    // Create Play Integrity API client
    const client = playintegrity({
      version: 'v1',
      auth: authClient,
    });

    // Decrypt the integrity token
    trackEvent('AndroidAuth.PlayIntegrity.DecodeStarted');
    const response = await client.v1.decodeIntegrityToken({
      packageName,
      requestBody: {
        integrityToken,
      },
    });
    const depDuration = Date.now() - depStart;

    if (!response.data || !response.data.tokenPayloadExternal) {
      trackDependency({
        name: 'PlayIntegrity.decodeIntegrityToken',
        target: 'playintegrity.googleapis.com',
        data: `packageName=${packageName}`,
        duration: depDuration,
        success: false,
        resultCode: 'empty-response',
        properties: { packageName },
      });
      return { ok: false, tolerable: false, reason: 'empty_response' };
    }

    // Extract and parse the tokenPayloadExternal
    const verdict = response.data.tokenPayloadExternal as PlayIntegrityVerdict;

    trackDependency({
      name: 'PlayIntegrity.decodeIntegrityToken',
      target: 'playintegrity.googleapis.com',
      data: `packageName=${packageName}`,
      duration: depDuration,
      success: true,
      resultCode: 200,
      properties: {
        packageName,
        appRecognition: verdict.appIntegrity?.appRecognitionVerdict,
        deviceRecognition: verdict.deviceIntegrity?.deviceRecognitionVerdict?.join(','),
        playProtect: verdict.environmentDetails?.playProtectVerdict,
      },
    });

    return { ok: true, verdict };
  } catch (error: unknown) {
    const depDuration = Date.now() - depStart;
    let status: number | string = 'exception';
    let message: string | undefined;

    // Handle Google API errors
    if (error && typeof error === 'object' && 'response' in error) {
      const apiError = error as {
        response?: {
          status?: number;
          statusText?: string;
          data?: unknown;
        };
        message?: string;
      };

      status = apiError.response?.status ?? 'exception';
      message = apiError.message;
    } else if (error instanceof Error) {
      message = error.message;
    }

    trackDependency({
      name: 'PlayIntegrity.decodeIntegrityToken',
      target: 'playintegrity.googleapis.com',
      data: `packageName=${packageName}`,
      duration: depDuration,
      success: false,
      resultCode: status,
      properties: { packageName, message },
    });
    trackException(new Error(message || 'Play Integrity API request failed'), {
      source: 'decryptAndVerifyIntegrityVerdict',
      packageName,
      status,
    });
    const unavailable = classifyUnavailable(error);
    if (unavailable) {
      return { ok: false, tolerable: true, reason: unavailable };
    }
    return { ok: false, tolerable: false, reason: 'api_error' };
  }
}
